waku/v2/protocol/app_design.go

```diff
@@ -0,0 +1,82 @@
+package protocol
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+type AppDesign struct {
+	mu      *sync.RWMutex
+	cancel  context.CancelFunc
+	ctx     context.Context
+	wg      *sync.WaitGroup
+	started bool
+}
+
+func NewAppDesign() AppDesign {
+	return AppDesign{
+		wg: &sync.WaitGroup{},
+		mu: &sync.RWMutex{},
+	}
+}
+
+func (sp *AppDesign) LockAndUnlock() func() {
+	sp.mu.Lock()
+	return func() {
+		sp.mu.Unlock()
+	}
+}
+func (sp *AppDesign) RLockAndUnlock() func() {
+	sp.mu.RLock()
+	return func() {
+		sp.mu.RUnlock()
+	}
+}
+
+// mutex protected start function
+// creates internal context over provided context and runs fn safely
+// fn is excerpt to be executed to start the protocol
+func (sp *AppDesign) Start(ctx context.Context, fn func() error) error {
```

Comment:
Let us add some unit tests to validate simultaneous start/stop, invoking stop without calling start etc.
Calling start and stop in a loop etc.